ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -301,19 +302,20 @@ protected void doOperationTowardsMaybeCompositeBlob(final String blobIdStr, Futu
       Callback<Void> callback, CompositeBlobOperationHelper helper, QuotaChargeCallback quotaChargeCallback) {
     // Can skip GET if we can determine this is not a metadata blob
     if (NonBlockingRouter.isMaybeMetadataBlob(blobIdStr)) {
-      Callback<GetBlobResultInternal> internalCallback = (GetBlobResultInternal result, Exception exception) -> {
+      Callback<GetBlobResult> internalCallback = (GetBlobResult result, Exception exception) -> {
         if (exception != null) {
           NonBlockingRouter.completeOperation(futureResult, callback, null, exception, false);
-        } else if (result.getBlobResult != null) {
+        } else if (result != null && result.getBlobDataChannel() != null) {
           exception = new RouterException(
               String.format("GET blob call returned the blob instead of just the store keys (before {} )",
                   helper.getOpName()), RouterErrorCode.UnexpectedInternalError);
           NonBlockingRouter.completeOperation(futureResult, callback, null, exception, false);
         } else {
           List<String> blobIdStrs = new ArrayList<>();
           blobIdStrs.add(blobIdStr);
-          if (result.storeKeys != null) {
-            result.storeKeys.forEach(key -> blobIdStrs.add(key.getID()));
+          List<StoreKey> blobChunkIds = result == null ? new ArrayList<>() : result.getBlobChunkIds();
+          if (blobChunkIds != null) {
+            blobChunkIds.forEach(key -> blobIdStrs.add(key.getID()));
```

Comment:
1> Why is result null ? A better way is for it to be never null and just be an empty list if we have nothing to serve

Reply:
Where is blobChunkIds used outside of this callback ? If not, why init it ?

Reply:
This is better.
```
if (result != null && result.getBlobChunkIds() != null) {
result.getBlobChunkIds().forEach(key -> blobIdStrs.add(key.getID()));
}
```

Reply:
If result.getBlobChunkIds() is never null, then you can skip that boolean

Reply:
Ok so one way to avoid this member BlobChunkIds from being null is to control its initialization in the constructor itself. This gets rid of all the null checks and eliminates any possiblity of null ptr exception. 

```
  public GetBlobResult(BlobInfo blobInfo, ReadableStreamChannel blobDataChannel, List<StoreKey> blobChunkIds) {
    this.blobInfo = blobInfo;
    this.blobDataChannel = blobDataChannel;
    this.blobChunkIds = (blobChunkIds == null) ? new ArrayList<>() : blobChunkIds;
  }
```

Reply:
Unless null is interpreted in a special way to mean something else elsewhere, I prefer to not have any nulls as much as possible.

Reply:
Fewer null checks in if-clauses makes code easier to read.

Reply:
I'd prefer result also to be not null, but that is beyond this PR

Reply:
blobChunkIds could be null when we GET without calling BlobChunkId subResource.
It's based on legacy code like how we handle GetBlobInternalResult. 
But I think you're right, I'm gonna have a follow up pr to set blobChunkIds as empty list if it don't have value at producer end. 
I would not prefer to handle it either in consumer or producer side instead of in the constructor.

Reply:
> I'm gonna have a follow up pr to set blobChunkIds as empty list if it don't have value at producer end.

```
    this.blobChunkIds = (blobChunkIds == null) ? new ArrayList<>() : blobChunkIds;
```

Its just a one line change in this pr. Follow up is not required.